### PR TITLE
FIX (GraphEngine): @W-12696959@: UnusedMethodRule now fully supports constructors.

### PR DESCRIPTION
**Pull request by jfeingold35**
This PR does the following:
- `UnusedMethodRule` now properly detects invocations of private/protected constructors via the `new` keyword (Thereby also resolving issue #1001)
- `UnusedMethodRule` now fully support public constructors

### CODE REVIEW

File: sfge/src/main/java/com/salesforce/rules/unusedmethod/RuleStateTracker.java [hunk 1]
```diff
@@ -238,6 +238,63 @@ boolean classInheritsMatchingMethod(String definingType, String signature) {
...
+                                                NodeType.NEW_OBJECT_EXPRESSION,
+                                                Schema.TYPE,
+                                                constructedType)));
+
+        // Inner types can be referenced by outer/sibling types by just the inner name rather than
+        // the full name.
+        // This means we need to do more, but what that "more" is depends on whether we're looking
+        // at an inner or outer type.
+        boolean typeIsInner = constructedType.contains(".");
+        if (typeIsInner) {
```
Comment (jfeingold35) [inline sfge/src/main/java/com/salesforce/rules/unusedmethod/RuleStateTracker.java:263]: The contents of this `if` branch are very similar to those of the one at line 316, but I couldn't see a way to cleanly refactor them into a single method. In the fullness of time, we may want to revisit that.

File: sfge/src/main/java/com/salesforce/rules/unusedmethod/RuleStateTracker.java [hunk 2]
```diff
@@ -371,4 +409,35 @@ Optional<BaseSFVertex> getDeclarationOfReferencedValue(MethodCallExpressionVerte
...
+    /**
+     * If an inner class shares the name of an outer class, then references to the inner class that
+     * occur in its outer/sibling classes can resemble references to the outer class. This class
+     * filters such false positives from the results of a query.
+     *
+     * @param unfilteredResults - The results, pre-filtering
+     * @param referencedType - The outer type with which inner types may collide
+     * @return - {@code unfilteredResults}, with all colliding references removed
+     */
+    private <T extends BaseSFVertex> List<T> removeInnerTypeCollisions(
```
Comment (jfeingold35) [inline sfge/src/main/java/com/salesforce/rules/unusedmethod/RuleStateTracker.java:422]: The else-branch from `getInvocationsOnType` and `getConstructionsOfType` are identical, and have been refactored into their own method.

File: sfge/src/test/java/com/salesforce/rules/unusedmethod/ConstructorsTest.java [hunk 3]
```diff
@@ -149,28 +137,50 @@ public void constructorUncalledByGrandchildSuper_expectViolation(Integer arity)
...
     }
 
+    /**
+     * Simple tests for cases where one class's constructor is called somewhere in another class.
+     *
+     * @param arity - The arity of the tested constructor.
+     */
+    @ValueSource(ints = {0, 1})
+    @ParameterizedTest(name = "{displayName}: Arity {0}")
+    public void constructorCalledViaNew_expectNoViolation(Integer arity) {
```
Comment (jfeingold35) [inline sfge/src/test/java/com/salesforce/rules/unusedmethod/ConstructorsTest.java:147]: New test for simply calling a constructor for one class in another class. This sort of thing is arguably covered by the tests for inner classes, but I felt that having an explicit test for it did no harm.

File: sfge/src/test/java/com/salesforce/rules/unusedmethod/ConstructorsTest.java [hunk 4]
```diff
@@ -149,28 +137,50 @@ public void constructorUncalledByGrandchildSuper_expectViolation(Integer arity)
...
-        "OuterClass.InnerClass, OuterClass.InnerClass",
-        "OuterClass.InnerClass, InnerClass",
-        "InnerClass, OuterClass.InnerClass",
-        "InnerClass, InnerClass"
-    })
-    @ParameterizedTest(name = "{displayName}: Declared as {0}, constructed as new {1}()")
-    @Disabled
-    public void innerConstructorUsedByOuter_expectNoViolation(String variable, String constructor) {
+    @ValueSource(strings = {"OuterClass.InnerClass", "InnerClass"})
+    @ParameterizedTest(name = "{displayName}: OuterClass.InnerClass constructed as new {0}()")
```
Comment (jfeingold35) [inline sfge/src/test/java/com/salesforce/rules/unusedmethod/ConstructorsTest.java:175]: The syntax with which the variable was declared ended up not impacting how the rule executes, so I removed the ability to configure that and cut the number of cases.

File: sfge/src/test/java/com/salesforce/rules/unusedmethod/ConstructorsTest.java [hunk 5]
```diff
@@ -184,19 +194,11 @@ public void innerConstructorUsedByOuter_expectNoViolation(String variable, Strin
...
-        "OuterClass.InnerClass, OuterClass.InnerClass",
-        "OuterClass.InnerClass, InnerClass",
-        "InnerClass, OuterClass.InnerClass",
-        "InnerClass, InnerClass"
-    })
-    @ParameterizedTest(name = "{displayName}: Declared as {0}, constructed as new {1}()")
-    @Disabled
-    public void innerConstructorUsedBySiblingInner_expectNoViolation(
-            String variable, String constructor) {
+    @ValueSource(strings = {"OuterClass.InnerClass", "InnerClass"})
```
Comment (jfeingold35) [inline sfge/src/test/java/com/salesforce/rules/unusedmethod/ConstructorsTest.java:199]: As above, variable declaration didn't terribly matter, so it was hardcoded.

File: sfge/src/test/java/com/salesforce/rules/unusedmethod/ConstructorsTest.java [hunk 6]
```diff
@@ -209,11 +211,54 @@ public void innerConstructorUsedBySiblingInner_expectNoViolation(
...
+     * These tests cover variants of a weird edge case: If an inner class and an outer class share
+     * the same name (e.g., {@code Whatever} and {@code Outer.Whatever}), then calling {@code new
+     * Whatever()} in the inner class's outer/sibling classes should count as an invocation for the
+     * inner class, not the outer class.
+     *
+     * @param arity - The arity of the tested constructor.
+     */
+    @ValueSource(ints = {0, 1})
+    @ParameterizedTest(name = "{displayName}: Constructor of arity {0}")
+    public void externalReferenceSyntaxCollision_expectViolation(Integer arity) {
```
Comment (jfeingold35) [inline sfge/src/test/java/com/salesforce/rules/unusedmethod/ConstructorsTest.java:232]: It seemed valuable to have tests for this kind of name ambiguity, so I added one.

File: sfge/src/test/java/com/salesforce/rules/unusedmethod/OverloadsTest.java [hunk 7]
```diff
@@ -319,7 +317,7 @@ public void callConstructorViaThisWithDifferentArityOverloads_expectViolation(
                         + String.format("    %s MyClass(boolean b, boolean c) {\n", scope)
                         + "    }\n"
                         // Use the engine directive to prevent this method from tripping the rule.
-                        + "    /* sfge-disable-stack UnusedMethodRule */"
+                        + "    /* sfge-disable-stack UnusedMethodRule */\n"
```
Comment (jfeingold35) [inline sfge/src/test/java/com/salesforce/rules/unusedmethod/OverloadsTest.java:320]: The missing newline meant that this directive was actually on the same line as the constructor and therefore wasn't being applied. Since public constructors weren't supported yet, that caused no problems. Adding support for public constructors caused the tests to start failing, and adding the newline fixes that.

File: sfge/src/main/java/com/salesforce/rules/unusedmethod/RuleStateTracker.java (hunk 1 above)
Comment (rmohan20) [reply sfge/src/main/java/com/salesforce/rules/unusedmethod/RuleStateTracker.java:263]: How about extracting a method and passing in `NodeType.<vertex_type>`, inner class name, and outer class name as input?
The name-splitting could become a util method in `ClassUtil`.

File: sfge/src/test/java/com/salesforce/rules/unusedmethod/ConstructorsTest.java (hunk 4 above)
Comment (rmohan20) [reply sfge/src/test/java/com/salesforce/rules/unusedmethod/ConstructorsTest.java:175]: How about we leave the initial set of ways of calling so that we'll know if there's ever a regression while making changes?

File: sfge/src/test/java/com/salesforce/rules/unusedmethod/ConstructorsTest.java (hunk 5 above)
Comment (rmohan20) [reply sfge/src/test/java/com/salesforce/rules/unusedmethod/ConstructorsTest.java:199]: Same as above. If we have a change in mechanism, the additional cases will be helpful to look for regressions.

**Review by rmohan20 — APPROVED** [review]
